client/client.go

```diff
@@ -1266,61 +1260,154 @@ func (c *Client) addAlloc(alloc *structs.Allocation) error {
+		c.heartbeatLock.Lock()
+
+		// If the last heartbeat didn't contain a leader, give the
+		// Nomad server this Agent is talking to one more attempt at
+		// providing a heartbeat that does contain a leader.
+		if atomic.LoadInt32(&c.lastHeartbeatFromQuorum) == 1 && now.Before(c.consulPullHeartbeatDeadline) {
+			c.heartbeatLock.Unlock()
+			// c.logger.Printf("[TRACE] client.consul: heartbeat received, sleeping until %v", c.consulPullHeartbeatDeadline)
+			return nil
+		}
+		c.heartbeatLock.Unlock()
+		c.logger.Printf("[TRACE] client.consul: lost heartbeat with Nomad quorum, falling back to Consul for server list")
 
-			c.configLock.RLock()
-			_, ok := c.configCopy.Node.Attributes["consul.server"]
-			c.configLock.RUnlock()
-			if !ok {
+		consulCatalog := c.consulSyncer.ConsulClient().Catalog()
+		dcs, err := consulCatalog.Datacenters()
+		if err != nil {
+			return fmt.Errorf("client.consul: unable to query Consul datacenters: %v", err)
+		}
+		dcs = dcs[0:lib.MinInt(len(dcs), datacenterQueryLimit)]
+		// Walk the list of Consul datacenters randomly in order to
+		// search for the Nomad server service.
+		shuffleStrings(dcs)
+
+		nomadServerServiceName := c.config.ConsulConfig.ServerServiceName
+		var mErr multierror.Error
+		const defaultMaxNumNomadServers = 8
+		nomadServerServices := make([]string, 0, defaultMaxNumNomadServers)
+		for _, dc := range dcs {
+			opts := &consulapi.QueryOptions{
+				AllowStale: true,
+				Datacenter: dc,
+				Near:       "_agent",
+				WaitTime:   consul.DefaultQueryWaitDuration,
+			}
+			consulServices, _, err := consulCatalog.Service(nomadServerServiceName, consul.ServiceTagRPC, opts)
+			if err != nil {
+				mErr.Errors = append(mErr.Errors, fmt.Errorf("unable to query service %q from Consul datacenter %q: %v", nomadServerServiceName, dc, err))
 				continue
 			}
-			services := make(map[string]struct{})
-			// Get the existing allocs
-			c.allocLock.RLock()
-			allocs := make([]*AllocRunner, 0, len(c.allocs))
-			for _, ar := range c.allocs {
-				allocs = append(allocs, ar)
+
+			for _, s := range consulServices {
+				port := strconv.FormatInt(int64(s.ServicePort), 10)
+				addr := s.ServiceAddress
+				if addr == "" {
+					addr = s.Address
+				}
+				serverAddr := net.JoinHostPort(addr, port)
+				serverEndpoint, err := rpcproxy.NewServerEndpoint(serverAddr)
+				if err != nil {
+					mErr.Errors = append(mErr.Errors, err)
+					continue
+				}
+				var ok bool
+				if ok, err = c.connPool.PingNomadServer(c.Region(), c.RPCMajorVersion(), serverEndpoint); err != nil {
+					mErr.Errors = append(mErr.Errors, err)
+					continue
+				}
+				if ok {
+					nomadServerServices = append(nomadServerServices, serverAddr)
+				}
+			}
+			// Break if at least one Nomad Server was successfully pinged
+			if len(nomadServerServices) > 0 {
+				break
+			}
+		}
+		if len(nomadServerServices) == 0 {
+			if len(mErr.Errors) > 0 {
+				return mErr.ErrorOrNil()
+			}
+
+			for i := range dcs {
```

Comment:
I explicitly did this so that each element is quoted: `"foo", "bar"` vs `foo, bar` because I wanted to make sure any spaces or non-ascii characters are quoted and escaped.  But, good point, I forgot the `+` format verb modifier `%+q`:

https://play.golang.org/p/7xkB1onztD